### PR TITLE
Feature/256 notification system

### PR DESCRIPTION
**Pull request by matir8**
## Description
- Added initial version of the notification system using RabbitMQ as a message broker. 
- Unit and integration tests are going to be added in a separate issue

## Type of change

-   [ ] Bug fix (non-breaking change which fixes an issue)
-   [x] New feature (non-breaking change which adds functionality)
-   [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
-   [x] This change requires a documentation update

## Checklist

-   [x] My code follows the style guidelines of this project
-   [x] I did not update API Controllers, if I did, I added/changed Postman tests
-   [x] I have performed a self-review of my own code
-   [x] I have commented my code, particularly in hard-to-understand areas
-   [ ] I have made corresponding changes to the documentation
-   [ ] I have added tests that prove my fix is effective or that my feature works
-   [x] New and existing unit tests pass locally with my changes
-   [x] I updated the changelog with an end-user readable description
-   [x] I assigned this pull request to the correct project board to update the sprint board

## Steps to Test or Reproduce

For a working instance of the notification service you should create a `.env` file in the root of the project with the following environmental variables:
```
RABBITMQ_HOST_NAME=rabbitmq    #if you run it from docker-compose use rabbitmq otherwise localhost for development environment
RABBITMQ_USERNAME=
RABBITMQ_PASSWORD=
SENDGRID_API_KEY=
SENDGRID_EMAIL_FROM=
```

In order to use the notification system in the API use the following snippet:
```
NotificationSender sender = new NotificationSender();
EmailNotification notification = new EmailNotification("test@example.com", "plain text content");
sender.RegisterNotification(Newtonsoft.Json.JsonConvert.SerializeObject(notification), Subject.EMAIL);
```

## Link to issue

Closes: #256

### CODE REVIEW

File: NotificationSystem/Services/EmailSender.cs [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+            client = new SendGridClient(Environment.GetEnvironmentVariable("SENDGRID_API_KEY"));
+            from = new EmailAddress(Environment.GetEnvironmentVariable("SENDGRID_EMAIL_FROM"));
+        }
+
+
+        public void SendNotification(INotification notification)
+        {
+            EmailNotification emailNotification = (EmailNotification) notification;
+
+            if (ValidateNotification(emailNotification))
```
Comment (waltersajtos) [inline NotificationSystem/Services/EmailSender.cs:28]: Might be an idea to throw an expectation when the notification is invalid

**Review by MeesvanStraten — DISMISSED** [review]
Got my email 🥇 , Code is solid and was able to get the email sent :)

**Review by niraymak — COMMENTED** [review]
Created an SendGrid account and added API key and was able to send plain HTML email! 👍 

I think we still need a general DeX SendGrid account. I was also wondering if the current implementation is able to send templates instead of plain HTML? I think this is a must for the near future but maybe for now plain HTML is OK. I will wait for Brend's feedback before approving.

**Review by Brend-Smits — CHANGES_REQUESTED** [review]
Appsettings should be used if possible, I will discuss this with you all in person

**Comment by Maxvanhattum** [comment]
Use these variables in the .env file for testing purposes, change sendgrid credentials if you want to send emails:

```
App__RabbitMQ__Hostname=rabbitmq #if you run it from docker-compose use rabbitmq otherwise localhost for development environment
App__RabbitMQ__Username=test
App__RabbitMQ__Password=test
App__SendGrid__ApiKey=test
App__SendGrid__EmailFrom=test@gmail.com
```

**Comment by Brend-Smits** [comment]
I've set up the SendGrid account and configured API keys for:
- Production
- Staging 
- Local development
All of these keys are stored in 1Password, everyone should have access to the Local Development key in there so they can test the email implementation when they want to. 

I'm gonna make some small changes to this code(https://github.com/DigitalExcellence/dex-backend/pull/322/files#diff-263557b791777f5b3ab458f4e289cd33f765c25fe77b1be8748397f09798e496R29-R46) later tonight and get the infrastructure ready and setup. 
I will keep you all posted!

**Review by MeesvanStraten — APPROVED** [review]

**Review by Brend-Smits — APPROVED** [review]

**Review by niraymak — APPROVED** [review]

**Pull request merged**